core/src/Cache/Keys/MsalAccountCacheKey.cs

```diff
@@ -26,30 +26,62 @@
 //------------------------------------------------------------------------------
 
 using System;
-using System.Collections.Generic;
-using System.Linq;
 using System.Text;
-using System.Threading.Tasks;
 
 namespace Microsoft.Identity.Core.Cache
 {
-    internal class MsalAccountCacheKey : MsalCacheKeyBase
+    /// <summary>
+    /// An object representing the key of the token cache Account dictionary. The 
+    /// format of the key is not important for this library, as long as it is unique.
+    /// </summary>
+    internal class MsalAccountCacheKey 
     {
-        public MsalAccountCacheKey(string environment, string tenantId, string userIdentifier) : base(environment, userIdentifier)
+        private readonly string _environment;
+        private readonly string _homeAccountId;
+        private readonly string _tenantId;
+
+        public MsalAccountCacheKey(string environment, string tenantId, string userIdentifier)
         {
-            TenantId = tenantId;
+            if (string.IsNullOrEmpty(environment))
+            {
+                throw new ArgumentNullException(nameof(environment));
+            }
+
+            _tenantId = tenantId;
+            _environment = environment;
+            _homeAccountId = userIdentifier;
         }
-        internal string TenantId { get; set; }
 
         public override string ToString()
         {
             var stringBuilder = new StringBuilder();
 
-            stringBuilder.Append(HomeAccountId + CacheKeyDelimiter);
-            stringBuilder.Append(Environment + CacheKeyDelimiter);
-            stringBuilder.Append(TenantId);
+            stringBuilder.Append(_homeAccountId + MsalCacheConstants.CacheKeyDelimiter);
+            stringBuilder.Append(_environment + MsalCacheConstants.CacheKeyDelimiter);
```

Comment:
Environment part of the key is now inconsistent between access token cache key and account cache key. 
Let's either only has scopes, or hash environment consistently.

Additional alternative for the fix - follow the documented cache key format everywhere, but hash the whole key only for UWP in platform specific code.

Reply:
Good point. I will hash only the scopes - this will give us a bit more flexibility in case we want to filter keys by env or client id etc. This only happen on UWP, there is no reason to change the logic for other platforms.

 In any case, since all the key components are also in the value, we can always write a one time migration code in case we need to make further changes. 

---
In reply to: [221720606](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1287#discussion_r221720606) [](ancestors = 221720606)

Reply:
@bgavrilMS Can you confirm that this change won't affect Android? And although, it only affects UWP, we should still document the change. Previous spec assumed same keys for all platforms unless described separately.

Reply:
@oldalton@microsoft.com - already updated the spec. 

---
In reply to: [222163772](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1287#discussion_r222163772) [](ancestors = 222163772)